include/mlir/Conversion/GPUToROCM/GPUToROCMPass.h

```diff
@@ -0,0 +1,102 @@
+//===- GPUToROCmPass.h - MLIR ROCm runtime support --------------*- C++ -*-===//
+//
+// Copyright 2019 The MLIR Authors.
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//   http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+// =============================================================================
+#ifndef MLIR_CONVERSION_GPUTOROCM_GPUTOROCMPASS_H_
+#define MLIR_CONVERSION_GPUTOROCM_GPUTOROCMPASS_H_
+
+#include <functional>
+#include <memory>
+#include <string>
+#include <vector>
+
+namespace mlir {
+
+namespace rocm {
+
+/// string constants used by the ROCM backend
+static constexpr const char *kHSACOAnnotation = "amdgpu.hsaco";
+static constexpr const char *kHSACOGetterAnnotation = "amdgpu.hsacogetter";
+static constexpr const char *kHSACOGetterSuffix = "_hsaco";
+static constexpr const char *kHSACOStorageSuffix = "_hsaco_cst";
+
+/// enum to represent the AMD GPU versions supported by the ROCM backend
+enum class AMDGPUVersion { GFX900 };
+
+/// enum to represent the HSA Code Object versions supported by the ROCM backend
+enum class HSACOVersion { V3 };
+
+/// Configurable parameters for generating the HSACO blobs from GPU Kernels
+struct HSACOGeneratorConfig {
+
+  /// Constructor - sets the default values for the configurable parameters
+  HSACOGeneratorConfig(bool isTestMode)
+      : testMode(isTestMode), amdgpuVersion(AMDGPUVersion::GFX900),
+        hsacoVersion(HSACOVersion::V3), rocdlDir("/opt/rocm/lib/"),
+        linkerPath("/opt/rocm/hcc/bin/ld.lld") {}
+
+  /// testMode == true will result in skipping the HASCO generation process, and
+  /// simply return the string "HSACO" as the HSACO blob
+  bool testMode;
+
+  /// the AMDGPU version for which to generate the HSACO
+  AMDGPUVersion amdgpuVersion;
+
+  /// the code object version for the generated HSACO
+  HSACOVersion hsacoVersion;
+
+  /// the directory containing the ROCDL bitcode libraries
+  std::string rocdlDir;
+
+  /// the path the ld.lld linker to use when generating the HSACO
+  std::string linkerPath;
+};
+
+} // namespace rocm
+
+using OwnedHSACO = std::unique_ptr<std::vector<char>>;
```

Comment:
Two reasons for this.

1. The binary blob owned by this variable could potentially be large, and hence pass-by-pointer instead of pass-by-value (Note that the "HSACO" string is only returned in the "test" scenario)

2. Consistency with the CUDA side (where we OwnedCubin, which is also unique_ptr wrapper on a vector of chars)

Reply:
> The binary blob owned by this variable could potentially be large, and hence pass-by-pointer instead of pass-by-value

A vector is very small (on the scale of 3 pointers / 24 bytes), you don't have to copy the heap allocated buffer.

> Consistency with the CUDA side (where we OwnedCubin, which is also unique_ptr wrapper on a vector of chars)

 That's a better reason, I should probably look into this.

Reply:
doesn't the  `std::vector<T>` own the memory used by its contents, and hence copying that vector could (sans compiler optimizations) also result in the need to allocate memory for its contents?

Reply:
If you pass it by value, you need to std::move it (same as the unique_ptr) to avoid the copy.
Just returning it by value already implies the move.